pkgs/build-support/rust/hooks/maturin-build-hook.sh

```diff
@@ -19,6 +19,7 @@ maturinBuildHook() {
         --target @rustTargetPlatformSpec@ \
         --manylinux off \
         --strip \
+        --interpreter $(command -v python) \
```

Comment:
I am not familiar enough with the Python machinery. @FRidh is `python` on `PATH` the Python of the build platform or host platform in cross-compilation?

@rmcgibbo I mean true cross-compilation here, not compilation under AArch64 emulation, which is strictly not cross-compilation (I think).

Reply:
`strictDeps` is used, so its the interpreter of the build platform.

Reply:
However, its probably better to hardcode the interpreter with a substitution, because there are alternative Python interpreters, such as `pypy`.

Reply:
> However, its probably better to hardcode the interpreter with a substitution, because there are alternative Python interpreters, such as `pypy`.

Yeah, that would be even better.

There are some variables that can be set for providing the correct header and library paths:

https://pyo3.rs/v0.13.2/building_and_distribution.html#cross-compiling

But that would only work for bindings that use PyO3. However, since PyO3 is the most commonly used create for binding Python, I guess the hook could detect if a package is using PyO3 and then set these variables in the expected way.

Reply:
> `strictDeps` is used, so its the interpreter of the build platform.

`maturin` does _run_ the interpreter in order to extract some metadata about it, so if it's the build platform's rather than the host platform's `python`, I'm not sure that's going to work without emulation...

> However, its probably better to hardcode the interpreter with a substitution, because there are alternative Python interpreters, such as pypy.

This was my first thought as well, but it wasn't clear to me how to accomplish it within the hook / nix code.

Reply:
> `maturin` does _run_ the interpreter in order to extract some metadata about it, so if it's the build platform's rather than the host platform's `python`, I'm not sure that's going to work without emulation...

The terminology here is a bit confusing in Nix, `buildPlatform` is the platform on which the package is built, `hostPlatform` is the platform you are building the package for. So, if it is the build platform, it is actually going to work without emulation...

...however, I am not sure that that would be good either, isn't the Python interpreter metadata supposed to be for the platform you are building for?

Reply:
* Thanks. I had that backward.
* This is the script they run: https://github.com/PyO3/maturin/blob/master/src/get_interpreter_metadata.py
* I suppose there's only one way to find out...
* n.b. are a bunch of recent tickets on the maturin issue tracker about cross-compilation as well, e.g https://github.com/PyO3/maturin/pull/454. We may be better off waiting for the dust to settle there?

Reply:
> ...however, I am not sure that that would be good either, isn't the Python interpreter metadata supposed to be for the platform you are building for?

A separate build interpreter is built. This interpreter adds the host metadata to sysconfig.

Reply:
In CPython cross-compilation does not really get much attention, so many downstreams such as Nixpkgs and Yocto have their patches.

Reply:
So, this will get messy if we use a substitution, because then the hook will depend on Python, and we need to rebuild the hook for each Python package set as well, resulting in a matrix of hooks. Not very nice.

Reply:
> So, this will get messy if we use a substitution, because then the hook will depend on Python, and we need to rebuild the hook for each Python package set as well, resulting in a matrix of hooks. Not very nice.

Agreed. It seems like it's going to be tricky to both:

  (a) have this work for pypy and other alternative python interpreters, and;
  (b) not need to rebuild the hook for each Python package set.
  
Unfortunately my PR against maturin (https://github.com/PyO3/maturin/pull/471) did not get a lot of traction, and filing a bug against rustlang or qemu (it's _probably_ a qemu bug, but I'm just guessing) is daunting.


I suppose I can just change this PR to something ugly like
```
get_python() {
    for name in "pypy" "python"; do
        path=$(command -v "$name" || echo "")
        if [ -f "$path" ]; then
            echo $path;
            break
        fi
    done
}

...

maturin ... \
  --interpreter $(get_python)
```

Reply:
maybe the interpreters should export a setup hook setting an alias `python`.